src/test/java/io/vavr/beanvalidation2/valueextraction/MapValueExtractorTest.java

```diff
@@ -0,0 +1,147 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.vavr.beanvalidation2.valueextraction;
+
+import io.vavr.collection.HashMap;
+import io.vavr.collection.Map;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.validation.ConstraintViolation;
+import javax.validation.Path;
+import javax.validation.Validation;
+import javax.validation.Validator;
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.Pattern;
+import java.util.Collection;
+import java.util.Iterator;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class MapValueExtractorTest {
+
+    private static final String SINGLE_CHAR = "^[a-z]$";
+    private Validator validator;
+
+    @Before
+    public void setUp() {
+        this.validator = Validation.buildDefaultValidatorFactory().getValidator();
+    }
+
+    private <T> void validateAndAssertNoViolations(T target) {
+        Collection<ConstraintViolation<T>> violations = validator.validate(target);
+        assertThat(violations).isEmpty();
+    }
+
+    private <T> void validateAndAssertSingleViolation(
+            T target, Class<?> constraint, String position, String type, String entry
+    ) {
+        Collection<ConstraintViolation<T>> violations = validator.validate(target);
+
+        assertThat(violations).isNotEmpty().hasSize(1);
+
+        ConstraintViolation<T> violation = violations.iterator().next();
+        assertThat(violation.getPropertyPath()).isNotEmpty().hasSize(2);
+
+        assertThat(violation.getConstraintDescriptor().getAnnotation()).isInstanceOf(constraint);
+        assertThat(violation.getPropertyPath().toString())
+                .isEqualTo("map" + type + "[" + position + "].<map " + entry + ">");
+
+        Iterator<Path.Node> iterator = violation.getPropertyPath().iterator();
+
+        Path.Node parent = iterator.next();
+        assertThat(parent.getName()).isEqualToIgnoringCase("map");
+
+        Path.Node child = iterator.next();
+        assertThat(child.getKey().toString()).isEqualToIgnoringCase(position);
+    }
+
+    @Test
+    public void defaultConstructionShouldValidate() {
+        validateAndAssertNoViolations(new TestBean());
+    }
+
+    @Test
+    public void havingEmptyValueShouldNotValidate() {
+        TestBean bean = new TestBean();
+        bean.put("b", "");
+        validateAndAssertSingleViolation(
+                bean, NotBlank.class, "b",
+                "<V>", "value"
```

Comment:
Java's Map produces here output: `map[b]<map value`, but Vavr's adds additional `<V>`. I don't know it that an issue or not.

Reply:
tbh, since we are using the interface correctly I think (of `ValueExtractor`) what the actual implementation (hibernate-validator) does with it - we should not care. that means we should not do string comparisons but rather query the `javax.validation.Path` via .getPropertyPath and not inspect the generated human readable string.

this way the tests could also be run against another implementation (non hibernate-validation). Can you rewrite it?

Reply:
I have mixed opinion on that. I agree, that we shouldn't rely on text representation of `javax.validation.Path`.

But this interface was designed for Java EE world when validation was done in back-end and validated values ware also presented in back-end. In that way one could rely on Java object representation of Path to properly notify on error, (in JSF it was automatic).

Nowadays validation results are mostly propagated to front-end client with some transport (mostly REST). Those transports are usually text based. Because of that people actually rely on output of `toString()` for Path as this is their only way of "serializing" validation violation across network.

Because of that, I've added those tests - to check the text representation that in fact matters for most JS/Java applications out there.

What do you think?

Reply:
Excellent question. But in my opinion this is not the responsibility of the validation provider to generate parseable path output here, rather the responsibility of the incorporating (REST-) framework such as spring, etc.

If bean validation is built-in those frameworks they are responsible for calling the provider and responding with a 400 providing additional information in a documented format. If not and you call the validation provider yourself, YOU are responsible to generate a 400 with information that you get back from the java API and make it "client-readable"

Bean validation should not care about REST, Change my view :)

Reply:
You're right that not responsibility of bean validation.

Unfortunately nobody (framework's authors) wants to gets their hands dirty and serialization of constraint violation is left to the users. I don't recall any framework actually do something better then call a `toString()` on Path. Also, most StackOverflow articles advice just that.

Let's leave it as it is. It's good enough.

Reply:
If the output were strictly json path or something I would say yeah, let's go for it, but it is not and vendor dependent. Maybe a good idea for a library on its own (bv-jsonpath)

I don't want the tests to break on library upgrade or implementation change. So I think we should cut dependency on tostring entirely

Reply:
Sure. I will rewrite those asserts to rely on on BV standard, so it will be implementation independent.